include/deal.II/matrix_free/tools.h

```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
       additional_data.mapping_update_flags_inner_faces |
       additional_data.mapping_update_flags_boundary_faces;
   }
-} // namespace MatrixFreeTools
 
+  namespace internal
+  {
+    template <typename Number>
+    struct LocalCSR
+    {
+      std::vector<unsigned int> row_lid_to_gid;
+      std::vector<unsigned int> row{0};
+      std::vector<unsigned int> col;
+      std::vector<Number>       val;
+    };
+  } // namespace internal
+
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components,
+            typename Number,
+            typename VectorizedArrayType>
+  void
+  compute_diagonal(
+    const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+    LinearAlgebra::distributed::Vector<Number> &        diagonal_global,
+    const std::function<void(FEEvaluation<dim,
+                                          fe_degree,
+                                          n_q_points_1d,
+                                          n_components,
+                                          Number,
+                                          VectorizedArrayType> &)> &local_vmult,
+    const unsigned int                                              dof_no,
+    const unsigned int                                              quad_no,
+    const unsigned int first_selected_component)
+  {
+    using VectorType = LinearAlgebra::distributed::Vector<Number>;
+
+    // initialize vector
+    matrix_free.initialize_dof_vector(diagonal_global);
+
+    int dummy;
+
+    matrix_free.template cell_loop<VectorType, int>(
+      [&](const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+          LinearAlgebra::distributed::Vector<Number> &        diagonal_global,
+          const LinearAlgebra::distributed::Vector<Number> &,
```

Comment:
shouldn't this argument be
```suggestion
          const int &,
```
to reflect the template argument?